File: Array/trapping-rain-water.cpp
//https://www.geeksforgeeks.org/trapping-rain-water/
// a[]=3   0   0   2   0   4
// l[]=3   3   3   3   3   4
// r[]=4   4   4   4   4   4
// w[]=3-3 3-0 3-0 3-2 3-0 4-4
// w = 0 + 3 + 3 + 1 + 3 + 0 = 10
#include <bits/stdc++.h>
using namespace std;

int main() 
{
	int t;
	cin>>t;
	while(t>0)
	{
	    int n;
	    cin>>n;
	    int arr[n];
	    for(int i=0;i<n;i++)
	    {
	        cin>>arr[i];
	    }
	    int left[n],right[n];
	    left[0] = arr[0];
	    right[n-1] = arr[n-1];
	    for(int i=1;i<n;i++)
	    {
	        left[i]=max(left[i-1],arr[i]);
	    }
	    for(int i=n-2;i>=0;i--)
	    {
	        right[i] = max(right[i+1],arr[i]);
	    }
	    int water =0;
	    for(int i=0;i<n;i++)
	    {
	        water = water + min(left[i],right[i])-arr[i];
	    }
	    cout<<water<<endl;
	    t--;
	}
	return 0;
}

File: Array/trapping-rain-water.py
# https://www.geeksforgeeks.org/trapping-rain-water/
# https://leetcode.com/problems/trapping-rain-water-ii/

'''
a=3, 0, 2, 0, 4
l=3, 3, 3, 3, 4
r=4, 4, 4, 4, 4


'''

def maxWater(a):
	n=len(a)
	if n == 0:	return 0
 
	left=[None]*n
	left[0]=a[0]
	for i in range(1,n):
		left[i]=max(left[i-1],a[i])
	
	right=[None]*n
	right[n-1]=a[n-1]
	for i in range(n-2,-1,-1):
		right[i]=max(right[i+1],a[i])
		
	water=0
	for i in range(n-1):
		water+=(min(left[i],right[i])-a[i])
	
	return water
	

if __name__=='__main__':
	a = [3, 0, 2, 0, 4]
	# a=[0,1,0,2,1,0,1,3,2,1,2,1]
	print("Max water={}".format(maxWater(a)))
